main.py

```diff
@@ -1 +1,121 @@
 # Secret Santa Chain Generator
+
+import csv
+import random
+import smtplib, ssl
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+from email.mime.image import MIMEImage
+
+# Extracts participants, emails, and exclusions from CSV file
+with open('assets/SS-participants.csv', newline='') as f:
+    reader = csv.reader(f)
+    participant_data = list(reader)
+
+# Extracts list of participants and emails from the data list
+participants = [i[1] for i in participant_data[1:]]
+names_drawn = []
+
+# Initialize seed
+random.seed(2019)
+
+# Generates a chain until it is valid
+isValid = False
+while not isValid:
+
+    # Creates duplicate list to modify
+    temp_participants = list(participants)
+    names_drawn = [i[:2] for i in participant_data[1:]]
+
+    isValid = True
+
+    # The first drawn name does not need to be checked for exclusions
+    rand = temp_participants[random.randint(0, len(temp_participants)) - 1]
+    first_drawn = rand
+    temp_participants.remove(rand)
+
+    for i in range(len(temp_participants)):
+
+        last_index = participants.index(rand)
+
+        # Create set of exclusions for the last drawn name
+        exclusions = set(participant_data[last_index + 1][2:])
+
+        # Create list of valid participants using the remaining names and the exclusion list
+        valid_participants = [item for item in temp_participants if item not in exclusions]
+
+        # The chain is invalid if there are no remaining valid participants
+        if not len(valid_participants):
+            isValid = False
+            break
+
+        else:
+
+            # Generates a random name from the remaining valid names
+            rand = valid_participants[random.randint(0, len(valid_participants)) - 1]
+
+            # Add the new generated name to the names drawn list next to the last generated name
+            names_drawn[last_index].append(rand)
+
+        # Add the valid name to the chain
+        temp_participants.remove(rand)
+
+    # The chain is invalid if the first name is in the exclusions list of the last name
+    last_index = participants.index(rand)
+    if first_drawn in participant_data[last_index + 1][1:]:
```

Comment:
I think this should only be executed on the last iteration, since this forces every participant not to have the first as an exclusion:

This order would be invalid:

1: Jeff exclusion Vanessa - Has Karl
2. Karl exclusion Jeff - Has Vanessa
3. Vanessa exclusion none - Has Jeff

The above would be invalid since `Jeff` is first drawn, then `Karl` we get down here and see that `Karl` has `Jeff` in his exclusion list aka he cannot draw `Jeff` next which he won't since `Jeff` has already been drawn and there are more people to be drawn ex. `vanessa`.

This if statment sees the 2nd link in the chain and invalidates the whole thing. Even though what and where we should check if Vanessa aka the last person.

^ Pretty sure this is a possible error that can come up. Lemme know if I just misread the code

Reply:
This happens after the for loop which loops through all the participants. The while loop keep running if the chain isn't valid. So its executed on the last iteration each time the while loop runs.